contrib/sbtc/token-key-admin/contracts/sbtc-token-key-admin.clar

```diff
@@ -0,0 +1,126 @@
+
+;; title: sbtc-token-key-admin
+;; version:
+;; summary:
+;; description:
+
+;; traits
+;;
+(impl-trait 'SP3FBR2AGK5H9QBDH3EEN6DF8EK8JY7RX8QJ5SVTE.sip-010-trait-ft-standard.sip-010-trait)
+
+;; token definitions
+;; 
+(define-fungible-token sbtc u2100000000000000)
+
+;; constants
+;;
+(define-constant contract-owner tx-sender)
+(define-constant err-invalid-caller u1)
+(define-constant err-owner-only u100)
+(define-constant err-not-token-owner u101)
+
+;; data vars
+;;
+(define-data-var coordinator (optional {addr: principal, key: (buff 33)}) none)
+
+;; data maps
+;;
+(define-map signers uint {addr: principal, key: (buff 33)})
+
+;; public functions
+;;
+(define-public (set-coordinator-data (data {addr: principal, key: (buff 33)}))
+    (if (is-valid-caller)
+        (ok (var-set coordinator (some data)))
+        (err err-invalid-caller)
+    )
+)
+
+(define-public (set-signer-data (id uint) (data {addr: principal, key: (buff 33)}))
+    (if (is-valid-caller)
+        (ok (map-set signers id data))
+        (err err-invalid-caller)
+    )
+)
+
+(define-public (delete-signer-data (id uint))
+    (if (is-valid-caller)
+        (ok (map-delete signers id))
+        (err err-invalid-caller)
+    )
+)
+
+(define-public (mint! (amount uint))
+    (if (is-valid-caller)
+        (token-credit! tx-sender amount)
+        (err err-invalid-caller)
+    )
+)
+
+(define-public (burn! (amount uint))
+    (if (is-valid-caller)
+        (token-debit! tx-sender amount)
+        (err err-invalid-caller)
+    )
+)
+
+(define-public (transfer (amount uint) (sender principal) (recipient principal) (memo (optional (buff 34))))
+	(begin
+		(asserts! (is-eq tx-sender sender) (err err-not-token-owner))
+		(try! (ft-transfer? sbtc amount sender recipient))
+		(match memo to-print (print to-print) 0x)
+		(ok true)
+	)
+)
+
+;; read only functions
+;;
+(define-read-only (get-coordinator-data)
+    (var-get coordinator)
+)
+
+(define-read-only (get-signer-data (signer uint))
+    (map-get? signers signer)
+)
+
+;;(define-read-only (get-signers)
+;;    (map-get? signers)
+;;)
+
+(define-read-only (get-name)
+	(ok "sBTC")
+)
+
+(define-read-only (get-symbol)
+	(ok "sBTC")
+)
+
+(define-read-only (get-decimals)
+	(ok u0)
+)
+
+(define-read-only (get-balance (who principal))
+	(ok (ft-get-balance sbtc who))
+)
+
+(define-read-only (get-total-supply)
+	(ok (ft-get-supply sbtc))
+)
+
+(define-read-only (get-token-uri)
+	(ok (some u"https://github.com/stacks-network/stacks-blockchain"))
+)
+
+;; private functions
+;;
+(define-private (is-valid-caller)
```

Comment:
Could we rename to this function `assert-tx-sender-is-owner!` and add `assert-tx-sender-is-coordinator!`?

Reply:
I named them ```is-contract-owner``` and ```is-coordinator``` since we might want to call them when we aren't doing an ```assert```.

Reply:
Makes sense. Could we have `assert-is-contract-owner` call `is-contract-owner`?

Reply:
Will do.

Reply:
I'm getting a weird ```clarinet check``` error when I try to add these functions.

`(define-private (assert-is-contract-owner)
    (asserts! (is-eq contract-owner tx-sender) (err err-invalid-caller))
)`

```detected two execution paths, returning two different expression types (got '(response UnknownType uint)' and 'bool')```

Reply:
The existing code which asserts `is-contract-owner` seems indistinguishable and doesn't error in the check

Reply:
Regarding the errors, you may need to add a begin block

Reply:
No, same error @igorsyl

Reply:
I think I understand; I'm guessing that ```asserts!``` in the true case returns true from the statement, and in the error case return err.  Since the compiler can't know which branch will be followed there is a conflict.

When using ```asserts!``` in the body of a function, though, the statement return from ```asserts!``` is ignored, but the function will return the error.

Reply:
Yeah that was it

Reply:
That doesn't work either:

 ;;
 (define-public (set-contract-owner (owner principal))
     (begin
-        (asserts! (is-contract-owner) (err err-invalid-caller))
+        (asserts-is-contract-owner!)
         (ok (var-set contract-owner owner))
     )
 )
@@ -173,3 +173,17 @@
 (define-private (is-valid-signer-id (id uint))
     (and (>= id u0) (< id (var-get num-keys)))
 )
+
+(define-private (asserts-is-contract-owner!)
+    (begin
+        (asserts! (is-contract-owner) (err err-invalid-caller))
+        (ok true)
+    )
+)

Still gives a clarity error:

```intermediary responses in consecutive statements must be checked```